test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java

```diff
@@ -0,0 +1,371 @@
+                readerGroupManager.getReaderGroup(readerGroupName).getGroupName(), readerGroupManager
+                        .getReaderGroup(readerGroupName).getScope(), readerGroupManager
+                        .getReaderGroup(readerGroupName).getOnlineReaders());
+        log.info("Creating {} readers", readers);
+        readerList = new ArrayList<>(readers);
+        log.info("Scope that is seen by readers {}", scope);
+
+        CompletableFuture.runAsync(() -> {
+            for (int i = 0; i < readers; i++) {
+                log.info("Starting reader: {}, with id: {}", i, readerName + i);
+                final EventStreamReader<Long> reader = clientFactory.createReader(readerName + i,
+                        readerGroupName,
+                        new JavaSerializer<Long>(),
+                        ReaderConfig.builder().build());
+                readerList.add(reader);
+                readerFutureList.add(startReading(reader));
+            }
+        });
+    }
+
+    void stopReadersAndWriters(ReaderGroupManager readerGroupManager, String readerGroupName) throws InterruptedException, ExecutionException {
+
+        log.info("Stop write flag status {}", testState.stopWriteFlag);
+        testState.stopWriteFlag.set(true);
+
+        log.info("Wait for writers execution to complete");
+        FutureHelpers.allOf(writerFutureList).get();
+        FutureHelpers.allOf(newWritersFutureList).get();
+        if (testState.getWriteException.get() != null) {
+            Assert.fail("Unable to write events. Test failure");
+        }
+
+        log.info("Stop read flag status {}", testState.stopReadFlag);
+        testState.stopReadFlag.set(true);
+
+        log.info("Wait for readers execution to complete");
+        FutureHelpers.allOf(readerFutureList).get();
+        if (testState.getReadException.get() != null) {
+            Assert.fail("Unable to read events. Test failure");
+        }
+
+        log.info("All writers have stopped. Setting stopReadFlag. Event Written Count:{}, Event Read " +
+                "Count: {}", testState.eventWriteCount.get(), testState.eventsReadFromPravega.size());
+        assertEquals(testState.eventWriteCount.get(), testState.eventsReadFromPravega.size());
+        assertEquals(testState.eventWriteCount.get(), new TreeSet<>(testState.eventsReadFromPravega).size()); //check unique events.
+
+        log.info("Closing writers");
+        writerList.forEach(writer -> writer.close());
+        newlyAddedWriterList.forEach(writer -> writer.close());
+        log.info("Closing readers");
+        readerList.forEach(reader -> reader.close());
+        log.info("Deleting readergroup {}", readerGroupName);
+        readerGroupManager.deleteReaderGroup(readerGroupName);
+    }
+
+
+    List<CompletableFuture<Void>> addNewWriters(String scope, ClientFactory clientFactory) {
+        //increase the number of writers to trigger scale
+        newlyAddedWriterList = new ArrayList<>(ADD_NUM_WRITERS);
```

Comment:
Per my earlier comment: newlyAddedWriterList should be a return value not a member variable.